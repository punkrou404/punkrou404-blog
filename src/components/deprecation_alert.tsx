import { Alert } from '@material-ui/lab';

const DAY = 1000 * 60 * 60 * 24;
const YEAR = DAY * 365;
const MONTH = YEAR / 12;

type DeprecationAlertProps = {
    date: string;
};

export const DeprecationAlert = ({ date }: DeprecationAlertProps): JSX.Element => {
    const current = new Date(date);
    const today = new Date();
    const diff = today.getTime() - current.getTime();
    const month = Math.floor(diff / MONTH);
    const year = Math.floor(diff / YEAR);

    if (year > 0) {
        return (
            <Alert severity="warning">{`この記事は最終更新日から${year}年以上が経過しています。`}</Alert>
        );
    }

    if (month > 0) {
        return (
            <Alert severity="warning">{`この記事は最終更新日から${month}か月以上が経過しています。`}</Alert>
        );
    }

    return <></>;
};
