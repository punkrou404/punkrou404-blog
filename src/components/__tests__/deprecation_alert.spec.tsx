import React from 'react';
import renderer from 'react-test-renderer';
import { DeprecationAlert } from '../deprecation_alert';

const ONE_DAY = 24 * 60 * 60 * 1000;
const ONE_MONTH = 32 * ONE_DAY;
const ONE_YEAR = 366 * ONE_DAY;

describe('DeprecationAlert', () => {
    it('today', () => {
        const beforeOneMonth = new Date().toISOString();
        const component = renderer.create(<DeprecationAlert date={beforeOneMonth} />);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('one month ago', () => {
        const beforeOneMonth = new Date(new Date().getTime() - ONE_MONTH).toISOString();
        const component = renderer.create(<DeprecationAlert date={beforeOneMonth} />);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('one year ago', () => {
        const beforeOneYear = new Date(new Date().getTime() - ONE_YEAR).toISOString();
        const component = renderer.create(<DeprecationAlert date={beforeOneYear} />);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });
});
