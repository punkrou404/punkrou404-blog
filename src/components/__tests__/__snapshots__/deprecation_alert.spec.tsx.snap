// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeprecationAlert one month ago 1`] = `
<div
  className="MuiPaper-root MuiAlert-root MuiAlert-standardWarning MuiPaper-elevation0"
  role="alert"
>
  <div
    className="MuiAlert-icon"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
      />
    </svg>
  </div>
  <div
    className="MuiAlert-message"
  >
    この記事は最終更新日から1か月以上が経過しています。
  </div>
</div>
`;

exports[`DeprecationAlert one year ago 1`] = `
<div
  className="MuiPaper-root MuiAlert-root MuiAlert-standardWarning MuiPaper-elevation0"
  role="alert"
>
  <div
    className="MuiAlert-icon"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
      />
    </svg>
  </div>
  <div
    className="MuiAlert-message"
  >
    この記事は最終更新日から1年以上が経過しています。
  </div>
</div>
`;

exports[`DeprecationAlert today 1`] = `null`;
